synapse/storage/databases/state/store.py

```diff
@@ -156,19 +167,50 @@ async def _get_state_groups_from_groups(
         """
         results = {}
 
-        chunks = [groups[i : i + 100] for i in range(0, len(groups), 100)]
-        for chunk in chunks:
-            res = await self.db_pool.runInteraction(
-                "_get_state_groups_from_groups",
-                self._get_state_groups_from_groups_txn,
-                chunk,
-                state_filter,
+        for group in groups:
+            results[group] = await self._get_state_groups_from_group(
+                group, state_filter
             )
-            results.update(res)
 
         return results
 
-    def _get_state_for_group_using_cache(self, cache, group, state_filter):
+    async def _get_state_groups_from_group(
+        self, group: int, state_filter: StateFilter
+    ) -> StateMap[str]:
+        """Returns the state groups for a given group from the
+        database, filtering on types of state events.
+
+        Args:
+            group: state group ID to query
+            state_filter: The state filter used to fetch state
+                from the database.
+        Returns:
+            state map
+        """
+
+        # convert the state_filter.types dict into something that is hashable.
+        frozen_kvs = {}
+        for k, v in state_filter.types.items():
+            if v is None:
+                frozen_kvs[k] = v
+            else:
+                # make the set hashable by making a frozen copy of it
+                frozen_kvs[k] = frozenset(v)
+
+        state_filter_hashable = (frozendict(frozen_kvs), state_filter.include_others)
```

Comment:
It takes a `Dict` in one of its attributes, which I could convert to a frozendict (though there are a lot of uses), if that seems like it'd be worth doing?

Reply:
I thought mostly things used one of the classmethods to construct `StateFIlter`? Otherwise you could use a `converter` to change the supplied dicts into frozen equivalents? Also heads up that `@attr.s(..)` has a `frozen` attribute

Reply:
A bit of grepping shows you're right — for that reason I didn't do the converter approach.